[Enhancement]: Provide logs up until the context deadline when failing to wait for condition

### Proposal

While I have seen this problem using the postgres module, I believe it applies to every Container Start failure with a WaitingFor strategy.

## Context

I have a postgres seed that is copied to the `/docker-entrypoint-initdb.d/seed.sql`, and this is used to prime the postgres DB at startup. It can happen that errors are introduced in the seed, causing the container to exit before the WaitingFor strategy is fulfilled. 

Here is an example failure:

```
psql:/docker-entrypoint-initdb.d/seed.sql:51: ERROR:  duplicate key value violates unique constraint "mytable_name_key"
DETAIL:  Key (name)=(a_name) already exists.
```

## Problem

It is hard to debug why the test fails because all output is swallowed and never output.

If the seed script contains an error, or violates the DB schema, the container will error out, and the Waiting condition will never fire. The failing container is removed so debugging the problem is made harder for the user.

## Proposition

If the [WaitingFor](https://github.com/testcontainers/testcontainers-go/blob/main/docker.go#L204-L209) condition fires with expired context, we should output what the container already logged:

```go
	if c.WaitingFor != nil {
		c.logger.Printf("🚧 Waiting for container id %s image: %s", shortID, c.Image)
		if err := c.WaitingFor.WaitUntilReady(ctx, c); err != nil {
			// We failed on the strategy, inform the user of logs up until now
			reader, err := c.Logs(ctx)
			if err != nil {
				return err
			}

			b, err := io.ReadAll(reader)
			if err != nil {
				return err
			}

			logs := string(b)
			fmt.Fprintf(os.Stderr, "failed waiting for condition: %s, container logs: %s", err, logs)
			return err
		}
	}
```

An example execution of a test failing because of the wait strategy would be:
```
2023/06/20 08:43:27 🐳 Creating container for image
2023/06/20 08:43:27 ✅ Container created: bc42707acea8
2023/06/20 08:43:27 🐳 Starting container: bc42707acea8
2023/06/20 08:43:28 ✅ Container started: bc42707acea8
2023/06/20 08:43:28 🚧 Waiting for container id bc42707acea8 image: fcd1ad29-af53-4055-9143-1b36c4fdea43:a13d99be-4bf6-4df3-adb9-08d78baacba7. Waiting for: &{timeout:<nil> Log:Ready to accept connections!!!! Occurrence:1 PollInterval:100ms}
    /Users/mdelapenya/sourcecode/src/github.com/testcontainers/testcontainers-go/docker_test.go:667: failed during started hook: container logs:
        1:C 20 Jun 2023 06:43:28.010 # oO0OoO0OoO0Oo Redis is starting oO0OoO0OoO0Oo
        1:C 20 Jun 2023 06:43:28.010 # Redis version=5.0.14, bits=64, commit=00000000, modified=0, pid=1, just started
        1:C 20 Jun 2023 06:43:28.010 # Warning: no config file specified, using the default config. In order to specify a config file use redis-server /path/to/redis.conf
        1:M 20 Jun 2023 06:43:28.011 * Running mode=standalone, port=6379.
        1:M 20 Jun 2023 06:43:28.011 # Server initialized
        1:M 20 Jun 2023 06:43:28.011 # WARNING you have Transparent Huge Pages (THP) support enabled in your kernel. This will create latency and memory usage issues with Redis. To fix this issue run the command 'echo never > /sys/kernel/mm/transparent_hugepage/enabled' as root, and add it to your /etc/rc.local in order to retain the setting after a reboot. Redis must be restarted after THP is disabled.
        1:M 20 Jun 2023 06:43:28.012 * Ready to accept connections

        : failed to start container
--- FAIL: TestContainerTerminationRemovesDockerImage (61.30s)
    --- FAIL: TestContainerTerminationRemovesDockerImage/if_built_from_Dockerfile (61.30s)
FAIL
FAIL	github.com/testcontainers/testcontainers-go	61.611s
```

Where the container logs are returned as part of the error, which is wrapped and eventually printed by the testing libraries

--- a/docker.go
+++ b/docker.go
@@ -193,22 +193,10 @@ func (c *DockerContainer) Start(ctx context.Context) error {
 		return err
 	}
 
-	shortID := c.ID[:12]
-
 	if err := c.provider.client.ContainerStart(ctx, c.ID, types.ContainerStartOptions{}); err != nil {
 		return err
 	}
 	defer c.provider.Close()
-
-	// if a Wait Strategy has been specified, wait before returning
-	if c.WaitingFor != nil {
-		c.logger.Printf("🚧 Waiting for container id %s image: %s", shortID, c.Image)
-		if err := c.WaitingFor.WaitUntilReady(ctx, c); err != nil {
-			return err
-		}
-	}
-
-	c.isRunning = true
 
 	err = c.startedHook(ctx)
 	if err != nil {
@@ -1026,6 +1014,24 @@ func (p *DockerProvider) CreateContainer(ctx context.Context, req ContainerReque
 							return fmt.Errorf("can't copy %s to container: %w", f.HostFilePath, err)
 						}
 					}
+
+					return nil
+				},
+			},
+			PostStarts: []ContainerHook{
+				// first post-start hook is to wait for the container to be ready
+				func(ctx context.Context, c Container) error {
+					dockerContainer := c.(*DockerContainer)
+
+					// if a Wait Strategy has been specified, wait before returning
+					if dockerContainer.WaitingFor != nil {
+						dockerContainer.logger.Printf("🚧 Waiting for container id %s image: %s", dockerContainer.ID[:12], dockerContainer.Image)
+						if err := dockerContainer.WaitingFor.WaitUntilReady(ctx, c); err != nil {
+							return err
+						}
+					}
+
+					dockerContainer.isRunning = true
 
 					return nil
 				},

--- a/docker.go
+++ b/docker.go
@@ -1025,7 +1025,10 @@ func (p *DockerProvider) CreateContainer(ctx context.Context, req ContainerReque
 
 					// if a Wait Strategy has been specified, wait before returning
 					if dockerContainer.WaitingFor != nil {
-						dockerContainer.logger.Printf("🚧 Waiting for container id %s image: %s", dockerContainer.ID[:12], dockerContainer.Image)
+						dockerContainer.logger.Printf(
+							"🚧 Waiting for container id %s image: %s. Waiting for: %+v",
+							dockerContainer.ID[:12], dockerContainer.Image, dockerContainer.WaitingFor,
+						)
 						if err := dockerContainer.WaitingFor.WaitUntilReady(ctx, c); err != nil {
 							return err
 						}

--- a/lifecycle.go
+++ b/lifecycle.go
@@ -2,6 +2,7 @@ package testcontainers
 
 import (
 	"context"
+	"io"
 	"strings"
 
 	"github.com/docker/docker/api/types/container"
@@ -130,6 +131,7 @@ func (c *DockerContainer) startingHook(ctx context.Context) error {
 	for _, lifecycleHooks := range c.lifecycleHooks {
 		err := containerHookFn(ctx, lifecycleHooks.PreStarts)(c)
 		if err != nil {
+			c.printLogs(ctx)
 			return err
 		}
 	}
@@ -142,11 +144,30 @@ func (c *DockerContainer) startedHook(ctx context.Context) error {
 	for _, lifecycleHooks := range c.lifecycleHooks {
 		err := containerHookFn(ctx, lifecycleHooks.PostStarts)(c)
 		if err != nil {
+			c.printLogs(ctx)
 			return err
 		}
 	}
 
 	return nil
+}
+
+// printLogs is a helper function that will print the logs of a Docker container
+// We are going to use this helper function to inform the user of the logs when an error occurs
+func (c *DockerContainer) printLogs(ctx context.Context) {
+	reader, err := c.Logs(ctx)
+	if err != nil {
+		c.logger.Printf("failed accessing container logs: %w\n", err)
+		return
+	}
+
+	b, err := io.ReadAll(reader)
+	if err != nil {
+		c.logger.Printf("failed reading container logs: %w\n", err)
+		return
+	}
+
+	c.logger.Printf("container logs:\n%s", b)
 }
 
 // stoppingHook is a hook that will be called before a container is stopped

--- a/lifecycle_test.go
+++ b/lifecycle_test.go
@@ -1,6 +1,7 @@
 package testcontainers
 
 import (
+	"bufio"
 	"context"
 	"fmt"
 	"strings"
@@ -14,6 +15,8 @@ import (
 	"github.com/docker/go-connections/nat"
 	"github.com/stretchr/testify/assert"
 	"github.com/stretchr/testify/require"
+	"github.com/testcontainers/testcontainers-go/internal/testcontainersdocker"
+	"github.com/testcontainers/testcontainers-go/wait"
 )
 
 func TestPreCreateModifierHook(t *testing.T) {
@@ -632,6 +635,76 @@ func TestLifecycleHooks_WithMultipleHooks(t *testing.T) {
 	require.Nil(t, err)
 
 	require.Equal(t, 20, len(dl.data))
+}
+
+type linesTestLogger struct {
+	data []string
+}
+
+func (l *linesTestLogger) Printf(format string, args ...interface{}) {
+	l.data = append(l.data, fmt.Sprintf(format, args...))
+}
+
+func TestPrintContainerLogsOnError(t *testing.T) {
+	ctx := context.Background()
+	client, err := testcontainersdocker.NewClient(ctx)
+	if err != nil {
+		t.Fatal(err)
+	}
+	defer client.Close()
+
+	req := ContainerRequest{
+		Image:      "docker.io/alpine",
+		Cmd:        []string{"echo", "-n", "I am expecting this"},
+		WaitingFor: wait.ForLog("I was expecting that").WithStartupTimeout(5 * time.Second),
+	}
+
+	arrayOfLinesLogger := linesTestLogger{
+		data: []string{},
+	}
+
+	container, err := GenericContainer(ctx, GenericContainerRequest{
+		ProviderType:     providerType,
+		ContainerRequest: req,
+		Logger:           &arrayOfLinesLogger,
+		Started:          true,
+	})
+	// it should fail because the waiting for condition is not met
+	if err == nil {
+		t.Fatal(err)
+	}
+	terminateContainerOnEnd(t, ctx, container)
+
+	containerLogs, err := container.Logs(ctx)
+	if err != nil {
+		t.Fatal(err)
+	}
+	defer containerLogs.Close()
+
+	// read container logs line by line, checking that each line is present in the stdout
+	rd := bufio.NewReader(containerLogs)
+	for {
+		line, err := rd.ReadString('\n')
+		if err != nil {
+			if err.Error() == "EOF" {
+				break
+			}
+
+			t.Fatal("Read Error:", err)
+		}
+
+		// the last line of the array should contain the line of interest,
+		// but we are checking all the lines to make sure that is present
+		found := false
+		for _, l := range arrayOfLinesLogger.data {
+			if strings.Contains(l, line) {
+				found = true
+				break
+			}
+		}
+		assert.True(t, found, "container log line not found in the output of the logger: %s", line)
+	}
+
 }
 
 func lifecycleHooksIsHonouredFn(t *testing.T, ctx context.Context, container Container, prints []string) {

